waiter/integration/waiter/websocket_integration_test.clj

```diff
@@ -192,6 +216,139 @@
         (finally
           (delete-token-and-assert waiter-url token))))))
 
+(deftest ^:parallel ^:integration-fast test-request-maintenance-mode
+  (testing-using-waiter-url
+    (let [ws-response-atom (atom [])
+          token (str "token-" (rand-name))
+          token-description (assoc (kitchen-request-headers :prefix "")
+                              :authentication "disabled"
+                              :maintenance {"message" "custom maintenance message"}
+                              :metric-group "waiter_ws_test"
+                              :name (rand-name)
+                              :permitted-user "*"
+                              :run-as-user (retrieve-username)
+                              :token token)
+          waiter-headers {"x-waiter-token" token}]
+      (testing "should not upgrade if token is in maintenance mode"
+        (try
+          (let [token-response (post-token waiter-url token-description)]
+            (assert-response-status token-response http-200-ok)
+            (try
+              (let [response-promise (promise)
+                    connection (ws-client/connect!
+                                 (websocket-client-factory)
+                                 (ws-url waiter-url "/websocket-auth")
+                                 (fn [{:keys [in out]}]
+                                   (async/go
+                                     (async/>! out "request-info")
+                                     (swap! ws-response-atom conj (async/<! in))
+                                     (swap! ws-response-atom conj (async/<! in))
+                                     (deliver response-promise :done)
+                                     (async/close! out)))
+                                 {:middleware (fn [_ ^UpgradeRequest request]
+                                                (websocket/add-headers-to-upgrade-request! request waiter-headers))})
+                    [close-code error] (connection->ctrl-data connection)]
+                (assert-websocket-upgrade-error-response close-code error  "Unexpected HTTP Response Status Code: 503 Service Unavailable"))))
+          (finally
+            (delete-token-and-assert waiter-url token))))
+
+      (testing "should give 400 if x-waiter-maintenance header is specified"
+        (try
+          (let [token-response (post-token waiter-url token-description)]
+            (assert-response-status token-response http-200-ok)
+            (try
+              (let [response-promise (promise)
+                    connection (ws-client/connect!
+                                 (websocket-client-factory)
+                                 (ws-url waiter-url "/websocket-auth")
+                                 (fn [{:keys [in out]}]
+                                   (async/go
+                                     (async/>! out "request-info")
```

Comment:
We expect the upgrade request to fail. This block should 
- close `out` immediately
- write to an atom and to trigger an assertion failure later when we check the `assert-websocket-upgrade-error-response`.